File: benchmarks/java/B3MatrixMul.java
public class B3MatrixMul {
    static final int SIZE = 2000;
    static long rngState = 42L;
    static double nextDouble() {
        rngState = rngState * 6364136223846793005L + 1442695040888963407L;
        return (double)(rngState >>> 33) / (double)(1L << 31);
    }
    public static void main(String[] args) {
        double[] A = new double[SIZE * SIZE], B = new double[SIZE * SIZE], C = new double[SIZE * SIZE];
        for (int i = 0; i < SIZE * SIZE; i++) { A[i] = nextDouble(); B[i] = nextDouble(); }
        for (int i = 0; i < SIZE; i++)
            for (int k = 0; k < SIZE; k++) {
                double aik = A[i * SIZE + k];
                for (int j = 0; j < SIZE; j++) C[i * SIZE + j] += aik * B[k * SIZE + j];
            }
        double sum = 0; for (double v : C) sum += v;
        System.out.printf("%.6f%n", sum);
    }
}

File: benchmarks/swift/b3_matrix_mul.swift
import Foundation
let SIZE = 2000

var rngState: UInt64 = 42
@inline(__always)
func nextDouble() -> Double {
    rngState = rngState &* 6364136223846793005 &+ 1442695040888963407
    return Double(rngState >> 33) / Double(UInt64(1) << 31)
}

// ContiguousArray + UnsafeBufferPointer: idiomatic Swift escape hatch for the
// hot loop. Plain [Double] keeps bounds checks and may dispatch through
// CoreFoundation bridging on every access — ~14x slower than this version
// without changing what the algorithm actually does.
var A = ContiguousArray<Double>(repeating: 0.0, count: SIZE*SIZE)
var B = ContiguousArray<Double>(repeating: 0.0, count: SIZE*SIZE)
var C = ContiguousArray<Double>(repeating: 0.0, count: SIZE*SIZE)

for i in 0..<SIZE*SIZE { A[i] = nextDouble(); B[i] = nextDouble() }

A.withUnsafeBufferPointer { aPtr in
    B.withUnsafeBufferPointer { bPtr in
        C.withUnsafeMutableBufferPointer { cPtr in
            for i in 0..<SIZE {
                for k in 0..<SIZE {
                    let a_ik = aPtr[i*SIZE+k]
                    for j in 0..<SIZE {
                        cPtr[i*SIZE+j] += a_ik * bPtr[k*SIZE+j]
                    }
                }
            }
        }
    }
}

var sum: Double = 0
for x in C { sum += x }
print(String(format: "%.6f", sum))
